FileNotFoundError for lock file with tmpdir for old pytest-of-user root

With an existing `/tmp/pytest-of-user` directory there will be an error now:

```
___ ERROR at setup of TestGeneralUsage.test_config_error ___

self = <CallInfo when='setup' exception: [Errno 2] No such file or directory: '/tmp/pytest-of-daniel/pytest-current/.lock'>, func = <function call_runtest_hook.<locals>.<lambda> at 0x7f88d52cc0d0>, when = 'setup'
treat_keyboard_interrupt_as_exception = False

    def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        self.when = when
        self.start = time()
        try:
>           self.result = func()

src/_pytest/runner.py:201: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/_pytest/runner.py:183: in <lambda>
    lambda: ihook(item=item, **kwds),
.venv/lib/python3.6/site-packages/pluggy/hooks.py:258: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
.venv/lib/python3.6/site-packages/pluggy/manager.py:67: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
.venv/lib/python3.6/site-packages/pluggy/manager.py:61: in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
src/_pytest/runner.py:104: in pytest_runtest_setup
    item.session._setupstate.prepare(item)
src/_pytest/runner.py:370: in prepare
    col.setup()
src/_pytest/python.py:1439: in setup
    fixtures.fillfixtures(self)
src/_pytest/fixtures.py:294: in fillfixtures
    request._fillfixtures()
src/_pytest/fixtures.py:467: in _fillfixtures
    item.funcargs[argname] = self.getfixturevalue(argname)
src/_pytest/fixtures.py:514: in getfixturevalue
    return self._get_active_fixturedef(argname).cached_result[0]
src/_pytest/fixtures.py:537: in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
src/_pytest/fixtures.py:623: in _compute_fixture_value
    fixturedef.execute(request=subrequest)
src/_pytest/fixtures.py:919: in execute
    return hook.pytest_fixture_setup(fixturedef=self, request=request)
.venv/lib/python3.6/site-packages/pluggy/hooks.py:258: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
.venv/lib/python3.6/site-packages/pluggy/manager.py:67: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
.venv/lib/python3.6/site-packages/pluggy/manager.py:61: in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
src/_pytest/fixtures.py:961: in pytest_fixture_setup
    result = call_fixture_func(fixturefunc, request, kwargs)
src/_pytest/fixtures.py:820: in call_fixture_func
    res = fixturefunc(**kwargs)
src/_pytest/pytester.py:371: in testdir
    return Testdir(request, tmpdir_factory)
src/_pytest/pytester.py:498: in __init__
    self.tmpdir = tmpdir_factory.mktemp(name, numbered=True)
src/_pytest/tmpdir.py:103: in mktemp
    return py.path.local(self._tmppath_factory.mktemp(basename, numbered).resolve())
src/_pytest/tmpdir.py:47: in mktemp
    p = make_numbered_dir(root=self.getbasetemp(), prefix=basename)
src/_pytest/tmpdir.py:66: in getbasetemp
    prefix="pytest-", root=rootdir, keep=3, lock_timeout=LOCK_TIMEOUT
src/_pytest/pathlib.py:233: in make_numbered_dir_with_cleanup
    consider_lock_dead_if_created_before=consider_lock_dead_if_created_before,
src/_pytest/pathlib.py:212: in cleanup_numbered_dir
    try_cleanup(path, consider_lock_dead_if_created_before)
src/_pytest/pathlib.py:194: in try_cleanup
    delete_a_numbered_dir(path)
src/_pytest/pathlib.py:166: in delete_a_numbered_dir
    create_cleanup_lock(path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

p = PosixPath('/tmp/pytest-of-user/pytest-current')

    def create_cleanup_lock(p):
        """crates a lock to prevent premature folder cleanup"""
        lock_path = get_lock_path(p)
        try:
>           fd = os.open(str(lock_path), os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644)
E           FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-user/pytest-current/.lock'

src/_pytest/pathlib.py:127: FileNotFoundError
============================================================================================== short test summary info ==============================================================================================
ERROR testing/acceptance_test.py::TestGeneralUsage::()::test_config_error
============================================================================================== 1 error in 0.32 seconds ==============================================================================================
―
```

This is due to recent changes, but have not bisected it.

--- a/changelog/4135.bugfix.rst
+++ b/changelog/4135.bugfix.rst
@@ -1,0 +1,1 @@
+pathlib based tmpdir cleanup now correctly handles symlinks in the folder.

--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -173,6 +173,8 @@ def delete_a_numbered_dir(path):
 
 def ensure_deletable(path, consider_lock_dead_if_created_before):
     """checks if a lock exists and breaks it if its considered dead"""
+    if path.is_symlink():
+        return False
     lock = get_lock_path(path)
     if not lock.exists():
         return True

--- a/testing/test_tmpdir.py
+++ b/testing/test_tmpdir.py
@@ -232,7 +232,7 @@ class TestNumberedDir(object):
 
         assert not lock.exists()
 
-    def test_cleanup_keep(self, tmp_path):
+    def _do_cleanup(self, tmp_path):
         self.test_make(tmp_path)
         from _pytest.pathlib import cleanup_numbered_dir
 
@@ -242,6 +242,9 @@ class TestNumberedDir(object):
             keep=2,
             consider_lock_dead_if_created_before=0,
         )
+
+    def test_cleanup_keep(self, tmp_path):
+        self._do_cleanup(tmp_path)
         a, b = tmp_path.iterdir()
         print(a, b)
 
@@ -275,3 +278,8 @@ class TestNumberedDir(object):
 
         rmtree(adir, force=True)
         assert not adir.exists()
+
+    def test_cleanup_symlink(self, tmp_path):
+        the_symlink = tmp_path / (self.PREFIX + "current")
+        the_symlink.symlink_to(tmp_path / (self.PREFIX + "5"))
+        self._do_cleanup(tmp_path)

--- a/testing/test_tmpdir.py
+++ b/testing/test_tmpdir.py
@@ -1,7 +1,10 @@
 from __future__ import absolute_import, division, print_function
 import sys
-import py
+
+import six
+
 import pytest
+from _pytest.pathlib import Path
 
 
 def test_tmpdir_fixture(testdir):
@@ -65,10 +68,6 @@ def test_basetemp(testdir):
     assert mytemp.join("hello").check()
 
 
-@pytest.mark.skipif(
-    not hasattr(py.path.local, "mksymlinkto"),
-    reason="symlink not available on this platform",
-)
 def test_tmpdir_always_is_realpath(testdir):
     # the reason why tmpdir should be a realpath is that
     # when you cd to it and do "os.getcwd()" you will anyway
@@ -78,7 +77,7 @@ def test_tmpdir_always_is_realpath(testdir):
     # os.environ["PWD"]
     realtemp = testdir.tmpdir.mkdir("myrealtemp")
     linktemp = testdir.tmpdir.join("symlinktemp")
-    linktemp.mksymlinkto(realtemp)
+    attempt_symlink_to(linktemp, str(realtemp))
     p = testdir.makepyfile(
         """
         def test_1(tmpdir):
@@ -281,5 +280,16 @@ class TestNumberedDir(object):
 
     def test_cleanup_symlink(self, tmp_path):
         the_symlink = tmp_path / (self.PREFIX + "current")
-        the_symlink.symlink_to(tmp_path / (self.PREFIX + "5"))
+        attempt_symlink_to(the_symlink, tmp_path / (self.PREFIX + "5"))
         self._do_cleanup(tmp_path)
+
+
+def attempt_symlink_to(path, to_path):
+    """Try to make a symlink from "path" to "to_path", skipping in case this platform
+    does not support it or we don't have sufficient privileges (common on Windows)."""
+    if sys.platform.startswith("win") and six.PY2:
+        pytest.skip("pathlib for some reason cannot make symlinks on Python 2")
+    try:
+        Path(path).symlink_to(Path(to_path))
+    except OSError:
+        pytest.skip("could not create symbolic link")

